### ISSUE TITLE
Document including the features/issues found in validation for the different runs

### ISSUE THREAD

**Issue by fjaviersanchez** [LSSTDESC/DC2-production#291]
I am opening this issue to track the progress in documenting the features and issues found in 1.1p, 1.2i/p and 2.0i. I will add a document (still to decide whether to be a link in the README in the Documents folder or an actual document there).

This is the summary of features/issues found during the validation:

* Run 1.1p: Issues with background (too much structure) created strange features in calexps. We thought that this prevented the input and output magnitudes to match. The issue had to do with optimization settings for the background (sky) photons. Extinction (both from the Milky Way and internal) was not applied to sources.

* Run 1.2p: Background now correct, this allowed us to find a mismatch between inputs and outputs. When we compared 1.2p outputs with 1.2i's we found that the number of detected sources was way larger in 1.2p than in 1.2i. After some digging, we found that extinction was not being applied in PhoSim (this was fixed in PhoSim after that). The PSF size/ellipticity tests show that the PSF ellipticity is slightly larger than *we think* it sould be (@rmandelb and @rmjarvis can elaborate more, also check #259). Photometry is correct if compared to reference catalog (see note 1) but biased if comparing with true catalog. Astrometry looks correct. Colors are wrong compared to protoDC2 due to bug while translating to instance catalogs.

* Run 1.2i: Good photometric and astrometric qualities (comparing to the reference catalog). PSF was too round/nice. Biased photometry if comparing to truth catalog (see note 1). Colors are wrong compared to protoDC2 due to the aforementioned but while translating to instance catalogs.

* Run 2.0i: Good photometric and astrometric qualities comparing to reference catalog (no truth catalog yet). Increased ellipticity in PSF. Colors are still wrong compared to cosmoDC2 (we found the bug while running 2.0).

* Note 1 (reference catalog vs truth catalog): The reference catalog was created directly from protoDC2/cosmoDC2 magnitudes and this was used to get the magnitude zeropoints and calibrate the sources. When we compare the measured magnitudes to the magnitude in these reference catalogs the distributions are centered at zero (except for the case of 1.1p), however when compared to the truth catalog they look biased. This is because when we translate the magnitudes from protoDC2/cosmoDC2 to the instance catalogs we find the best fit SED in the CatSim library, however, the normalization chosen seems to produce some mismatch in color space (more details on this in @danielsf's DC2 presentations). Several solutions proposed and there's ongoing work to mitigate this.

Some other details if you are interested in the full story behind DC2 (DC2 collector's-edition):

* Note 2 (reprocessings): After the first processing of 1.2i, we noticed that there were some astrometric and photometric biases, this turned out to be caused by the galaxies entering the reference catalogs. After that, we changed the settings for processing and reprocessed both 1.2i and 1.2p. The astrometry and photometry biases disappeared (comparing to the reference).

* Check #259, #278 for validation in 1.2i/p and 2.0i. Also check the notebooks [here](https://github.com/LSSTDESC/DC2-analysis/tree/validation/validation) if you want to take a look at some results/use these pieces of code for your analyses.

**Comment by yymao** [comment]
This note is great! Thanks, Javier!

**Comment by rmjarvis** [comment]
I think this is all correct.  Thanks, Javi, for documenting it.

**Comment by boutigny** [comment]
Thanks Javier, this is very useful. 
I have one comment regarding the astrometry: I don't think that we changed anything to use only stars, I don't think that it is even possible at the configuration level. A change has been applied for the photometry, this is this entry in the config file: `config.calibrate.photoCal.match.sourceSelection.unresolved.maximum=0.5` 
But for the astrometry I am pretty sure that we are still using stars and galaxies from the reference catalog. The only change which has been done by @danielsf  at an early stage of the processing  was to limit the magnitude of the objects in the reference catalog to 23.

**Comment by fjaviersanchez** [comment]
@boutigny thanks a lot for the clarification!

**Comment by cwwalter** [comment]
Is there a reference for: "Colors are wrong compared to protoDC2 due to bug while translating to instance catalogs." so we can document it?

**Comment by fjaviersanchez** [comment]
I found [this presentation](https://docs.google.com/presentation/d/1DNu1bX1Mpldvn8PY6l3QIkVkKueXOV1xbE5gaowsG4M/edit#slide=id.g3f6d327da3_0_5) and [this other presentation](https://docs.google.com/presentation/d/173ihFjV1hhLKUPQK0gdxXtMjEiRmExpOrL094Khd0ow/edit#slide=id.g44596d4d82_15_52). I think there's more documentation elsewhere but I can't remember where, @danielsf may have something.

**Comment by cwwalter** [comment]
Oh... I was thinking you were saying that this was a different problem that we solved with the top hat since it is listed in the 1.2 section.

I remembered we determined that with 2.0. But I missed your "Colors are still wrong compared to cosmoDC2 (we found the bug while running 2.0)." So got it now.  Thanks.

**Comment by fjaviersanchez** [comment]
Here's a great notebook by @plaszczy illustrating some of the features in 1.2p [here](https://github.com/LSSTDESC/DC2-production/blob/u/plaszczy/nb_run1_2/Notebooks/footprint_Run12p_spark.ipynb). Thanks a lot for putting this together.

**Comment by fjaviersanchez** [comment]
Update (9/12): Preliminary tests on Run2.1i

Visit 1918527 r-band, Sky brightness: 20.43896 (r-band), Moon: 33 deg, Phase: 69% (100% full Moon)

Density: 10.2 galaxies/sq-arcmin. Depth: 23.5

Measured magnitude for objects used for photometric calibration (`calib_photometry_used==True`) and matched object in the reference catalog: 

![image](https://user-images.githubusercontent.com/13971732/49707273-75f3f080-fbdf-11e8-9bf5-36cd1cbbf6da.png)

Kron photometry (all objects):

![image](https://user-images.githubusercontent.com/13971732/49707376-e7cc3a00-fbdf-11e8-976a-fec974f96d88.png)

Astrometric residuals for objects with `calib_astrometry_used==True`:

![image](https://user-images.githubusercontent.com/13971732/49707409-03374500-fbe0-11e8-808c-e99663c49ddc.png)

Astrometry (all objects):

![image](https://user-images.githubusercontent.com/13971732/49707431-1a763280-fbe0-11e8-9a64-bf0040a5bf01.png)

Detection efficiency for stars with |mag_PSF - mag_true| < 0.1 and detected within 5 pixels of their true centroid:

![image](https://user-images.githubusercontent.com/13971732/49707482-48f40d80-fbe0-11e8-9505-c0f0bb564ca3.png)

PSF size:

![image](https://user-images.githubusercontent.com/13971732/49707506-5b6e4700-fbe0-11e8-820a-63d3edb8f96d.png)

Checking for brighter-fatter:

![image](https://user-images.githubusercontent.com/13971732/49707527-6a54f980-fbe0-11e8-9da8-6d63cd2cf97f.png)

Depth distribution across the focal plane:

![image](https://user-images.githubusercontent.com/13971732/49707540-7a6cd900-fbe0-11e8-8cc6-ed1c931e9298.png)

![image](https://user-images.githubusercontent.com/13971732/49707556-835daa80-fbe0-11e8-884f-bc5a04c7318a.png)

PSF ellipticity distribution across the focal plane (median and 95-th percentile marked with broken lines):

![image](https://user-images.githubusercontent.com/13971732/49707563-983a3e00-fbe0-11e8-93da-d2fa4053e8b5.png)

I have to repeat these checks with the results for sip order = 2

**Comment by rmjarvis** [comment]
> Astrometric residuals for objects with calib_astrometry_used==True:

Question about this test: Are you applying the cos(dec) factor to the RA residuals?  

The distance on the sky between two objects with the same Dec, but a small Delta RA is (Delta RA) * cos(Dec).  So if you are currently just reporting the Delta RA value, then it's probably worth multiplying by cos(Dec) to turn this into an angle on the sky.  Otherwise, exposures closer to the poles will seem to have worse astrometry in the RA direction, when it's just a measurement artifact of using RA, Dec coordinates.

**Comment by fjaviersanchez** [comment]
Thanks a @rmjarvis. I wasn't including this. However, I thought I needed to multiply times sin(dec)?

This is what I get using sin(dec):

![image](https://user-images.githubusercontent.com/13971732/49708744-b8b8c700-fbe5-11e8-8151-875b00fca593.png)

And this is using cos(dec):

![image](https://user-images.githubusercontent.com/13971732/49708772-dab24980-fbe5-11e8-9486-205893343932.png)

**Comment by rmjarvis** [comment]
No, it should be cos(dec).  At the equator, there is no effect.  At the poles, small angles have large delta RA.  

It's certainly possible that the astrometry residuals aren't the same in RA and Dec.  So I don't think it's indicative of a problem that the blue and orange aren't precisely on top of each other even after applying the cos(dec) factor.

**Comment by fjaviersanchez** [comment]
Update 12/10:

I also checked the u-band pre-2.1i images and they look good to me:

Number density (~6.8 galaxies/sq-arcmin). Photometry looks good, astrometry is almost unbiased (RA shows a median difference of 1.5 mas with respect to the reference catalog but this is probably due to statistics and the matching method).

I am showing the same plots as above:

Photometry for calibration objects
![image](https://user-images.githubusercontent.com/13971732/49774128-9045d100-fca8-11e8-9ebc-99dfa7c2e797.png)
As a function of magnitude
![image](https://user-images.githubusercontent.com/13971732/49774142-9b006600-fca8-11e8-946d-8274368c45d4.png)
Astrometry for calibration objects
![image](https://user-images.githubusercontent.com/13971732/49774170-b23f5380-fca8-11e8-8a9a-d2b7b7d1b505.png)
(the plots above and below contains the factor cos(dec))
For all objects
![image](https://user-images.githubusercontent.com/13971732/49774246-f3376800-fca8-11e8-8336-608f84c261bb.png)
PSF size:

![image](https://user-images.githubusercontent.com/13971732/49774287-19f59e80-fca9-11e8-8314-50fd7f771e5d.png)

Looking for BF effect:

![image](https://user-images.githubusercontent.com/13971732/49774299-2679f700-fca9-11e8-9089-154798af6e02.png)

Detection efficiency for stars:

![image](https://user-images.githubusercontent.com/13971732/49774318-31348c00-fca9-11e8-86ef-95a5201a7a40.png)

Depth:

![image](https://user-images.githubusercontent.com/13971732/49774358-46a9b600-fca9-11e8-9322-9d953481dd16.png)

![image](https://user-images.githubusercontent.com/13971732/49774383-78228180-fca9-11e8-9ede-b8e34b9d7b72.png)

PSF ellipticity:

![image](https://user-images.githubusercontent.com/13971732/49773842-90919c80-fca7-11e8-9620-e759ecd5c049.png)

One of the critical things to check are the interactions between the different types of objects (as rendered by imSim). I'll keep working on this.

**Comment by fjaviersanchez** [comment]
Update 12/12:

I started checking the centroid file. I matched the objects in the output catalog to the closest neighbor in the centroid file with flags!=1 (which means skipped). This is a very crude approximation since the centroid files include more than one component, however, I assumed that the dominant component in the object will be the one that its centroid lies closest to the measured centroid.

With that in mind, I tried to check some quantities classifying by flags. The flags are the following:

0 -> No flags
2 -> Simple SED
4 -> No silicon
8 -> FFT rendered

If an object has more than one flag, they are added up, e.g, an object that doesn't use the silicon model and has a simple sed has flags=6. An object that has been FFT rendered and doesn't use the silicon model has flags=12, and so on (there are very few of these because these are the brightest).

I checked the magnitude distribution for the available r-band visit:

![image](https://user-images.githubusercontent.com/13971732/49910402-738bd380-fe37-11e8-89f5-c6772493ddb2.png)

This looks reasonable to me, the objects with flags=0 or 2 are overall brighter than those that use flags=6 (simple_sed+no_silicon). 

I also checked the ellipticity distributions and they all look correct:

![image](https://user-images.githubusercontent.com/13971732/49910449-a1711800-fe37-11e8-8e33-1f31345c1726.png)

![image](https://user-images.githubusercontent.com/13971732/49910463-a7ff8f80-fe37-11e8-8aa1-f262b11848f9.png)

![image](https://user-images.githubusercontent.com/13971732/49910468-ae8e0700-fe37-11e8-9abd-6520b91cd854.png)

These are the HSM PSF-corrected regauss e1,e2 from the stack.
@rmjarvis @cwwalter is there anything else you'd like to see?

I also tried to match every truth component but the histograms were dominated by the "by-chance" matches (since the overall density in the instance catalogs is way higher than the density in the output single-visit catalogs). I can try some smarter ways to match in order to get more accurate results as well but I think this was a good quick and dirty approximation to get an overall view. Comments and requests are welcome 🙂

**Comment by cwwalter** [comment]
Thanks Javier!

Can you explain the magnitude plot more?

Is the y-axis a fractional amount of something?  I can't understand why 

- The numbers of objects are the same for all classes
- The distribution for "normal" goes all the way down to 26 and the distribution for dim simple sed objects goes all the way out to 16.  I must not be understanding something about the plotting...

A couple of other ideas:

- The objects IDs are actually combined bit patters with the lower 10 bits being the object type and the upper bits being the uniqueID.  So using this info you can actually combine objects if you want.

- Also it would be interesting to make a "truth" plot. Using only the centroid file you can plot the number of each class as a function of magnitude (including skipped ones).

**Comment by cwwalter** [comment]
For the plots from yesterday, is the BF as large as we saw it before?  (There was a small change in the GalSim 2.1's diffusion model so would like to see there was no big change).

**Comment by fjaviersanchez** [comment]
> Is the y-axis a fractional amount of something? I can't understand why

Sorry, I should have specified that. I normalized all the histograms to check that the distributions had reasonable shapes.

> * The distribution for "normal" goes all the way down to 26 and the distribution for dim simple sed objects goes all the way out to 16.  I must not be understanding something about the plotting...

The objects at 26 may be bad matches, I can check the differences in total flux as well just to make sure. I would say also the same for the objects at mag=16. I assume that a good fraction of them should be faint objects that aren't detected and overlap with some brighter object but I have to check that.


> * The objects IDs are actually combined bit patters with the lower 10 bits being the object type and the upper bits being the uniqueID.  So using this info you can actually combine objects if you want.
> *

I was thinking of using these but the flags may be different for the different components and I didn't know which approach would be better (I can probably add and/or multiply the flags of each component to get a combined flag)


> * Also it would be interesting to make a "truth" plot. Using only the centroid file you can plot the number of each class as a function of magnitude (including skipped ones)

Thanks! That's a good idea!



> For the plots from yesterday, is the BF as large as we saw it before? (There was a small change in the GalSim 2.1's diffusion model so would like to see there was no big change).

I think that these same exact visits are not available for run 2.0i. I would say that the effect is smaller than in run 1.2i though (the PSF for that particular visit was slightly larger but comparable):

![image](https://user-images.githubusercontent.com/13971732/49914547-8a85f200-fe46-11e8-8fa5-8dba37457770.png)

**Comment by fjaviersanchez** [comment]
To be more concise.

Run 1.2i:

![image](https://user-images.githubusercontent.com/13971732/49914621-e3558a80-fe46-11e8-80e7-b67590bfed9a.png)

![image](https://user-images.githubusercontent.com/13971732/49914626-e6e91180-fe46-11e8-97b8-15cffde7344c.png)

Run 2.1i

![image](https://user-images.githubusercontent.com/13971732/49914634-f10b1000-fe46-11e8-9715-c9396c0269e7.png)

![image](https://user-images.githubusercontent.com/13971732/49914645-f8321e00-fe46-11e8-9849-5477ebef89f2.png)

**Comment by cwwalter** [comment]
Yes, that's what i was afraid of.  It looks like BF is missing right? Or am I comparing the wrong plots? I see it clearly in 1.2 and not at all in 2.0...

**Comment by cwwalter** [comment]
> Sorry, I should have specified that. I normalized all the histograms to check that the distributions had reasonable shapes.

OK thanks. I'm sort of surprised.  Maybe the matching isn't really working but these distributions basically look all the same for all the different classes. I would have expected all the dim ones to be on the low end.  The "truth" plot should show that I think. I don't expect to tag as many "dim" objects at 20 as 24.

**Comment by rmjarvis** [comment]
Hm.  That B/F plot does look like a problem.  The other thing that doesn't seem right is the magnitude histogram.  The flag=2 objects should all be at mag > ~26 I believe.  Could most of these be bad matches?  (Similar point to what Chris just posted.)

**Comment by jchiang87** [comment]
> @jchiang87 can we confirm that the flags for the test were correct and --disable_sensor_model wasn't turned on or something?

No way was `--disable_sensor_model` turned on.  These data have tree rings.  As for confirming the flags are correct:  There should be enough info in the centroid files to determine if the reported flags are at least consistent, i.e., flag=12 will have large `Flux` values, etc.; flag=2 or 6 should have small `Flux` values.  Checking one of the files, this seems to be the case.   Plus, the code looks correct.

**Comment by jchiang87** [comment]
It seems to me that if you want to determine the distribution of object magnitudes (or fluxes) as a function of the rendering flag value, you can do that with just the centroid file, i.e., you don't need to match against an external catalog or use any outputs from the DM stack processing.

**Comment by cwwalter** [comment]
OK that's what I figured.. the plot showing the categories might just be bad matching.  The truth plot will show us what we expect but it sounds like you already peaked at files and it looks reasonable.

If that is all true; the BF is harder to understand. There was this diffusion change but (assuming it is correct) I don't think this should be a large effect.

Javier, how exactly are you calculation deltaT again?  Maybe it is a measurement issue?

**Comment by cwwalter** [comment]
> It seems to me that if you want to determine the distribution of object magnitudes (or fluxes) as a function of the rendering flag value, you can do that with just the centroid file, i.e., you don't need to match against an external catalog or use any outputs from the DM stack processing.

Right... that is what I am talking about with the "truth plot".  Just use the centroid file itself.

**Comment by cwwalter** [comment]
Nice!

That's more like I expected.  In case people are wondering, the reason you see stuff down at 40 is that those are dim galaxy components.Disks and bulges and knots(?) are entered separately  Our cut is on the combined galaxy magnitude.

**Comment by cwwalter** [comment]
And, in the BF region all "normally handled" objects.. 🤔

**Comment by cwwalter** [comment]
@rmjarvis Does the GalSim test_sensor unit test check if BF is working as expected?

**Comment by fjaviersanchez** [comment]
> Javier, how exactly are you calculating deltaT again? Maybe it is a measurement issue?

I compute `T= Ixx+Iyy` and `deltaT = T - average(T[(mag>20) & (mag<22)])`.

Thanks for the plots @jchiang87, I think the way you did it is probably the most relevant to check that these approximations worked properly. However, I wanted to check that this was translated correctly into the images and I think that I'll have to do something more clever with the matching to check that. In any case, I think I am overall happy with the results (modulo BF).

**Comment by rmjarvis** [comment]
> Does the GalSim test_sensor unit test check if BF is working as expected?

Yes.  To a point, but it does check that stars get bigger with the silicon sensor for high fluxes.  Those all still pass after Craig's update.  And he rand more involved validations that tested the slope of the effect to higher accuracy than the unit tests check it.

**Comment by jchiang87** [comment]
The shift to higher magnitudes is just a consequence of using different gains for the two datasets.  In Run2.1i, we have gain = 0.7 e-/adu, while for Run2.0i, we had gain = 1.7 e-/adu.  That gives a shift in magnitude of -2.5*log10(0.7/1.7) = 0.96.

**Comment by jchiang87** [comment]
Thinking about it some more, that gain change explanation for the shift in the saturation onset can't be right.  The zero points also shift by that same amount, so the saturation should kick in at the same magnitude, which makes sense since an object's magnitude shouldn't depend on the gain either.

**Comment by rmjarvis** [comment]
How are we defining saturation?  In terms of electrons or ADU?  It should be based on the number of electrons in a pixel.  But maybe the saturation is defined in terms of ADU, and we didn't change that when we changed the gain?

**Comment by RobertLuptonTheGood** [comment]
We quote saturation levels in ADU, not electrons, as we expect that the gain is more stable than our knowledge of it.

					R

**Comment by jchiang87** [comment]
In the simulation, the saturation level is set at 1e5 e-/pixel.

**Comment by fjaviersanchez** [comment]
The objects with `flags==12` seem to actually be the brightest in the exposure. Here I am showing an image with markers on all objects with `flags==12`:

![image](https://user-images.githubusercontent.com/13971732/49987596-aa8de200-ff28-11e8-8ee4-43acec54173d.png)

I am having some trouble using `base_SdssShape_x/y` to match to x,y in the centroid files though. I'll follow up tomorrow.

**Comment by cwwalter** [comment]
When I was thinking about this yesterday, I looked at Jim's plot above and saw the flagid=12 guys all had magnitudes of less than 15.  But we see the saturation kicking in earlier than that.  So, I was thinking that couldn't be it. 

I wonder if I either misinterpreted the plot or if it is possible something about the FFT PSF for the very brightest objects is causing the photometry to measure them as dimmer objects.

**Comment by rmjarvis** [comment]
I can't think of any way that would happen.  Javi mentioned a brighter sky background, but that also doesn't make much sense, since it would have to be **much** brighter for it to move the saturation point a full magnitude.

I think the best bet is probably to look at the images of an object that is saturated in 2.1 and not in 2.0.  Maybe that will give a clue.  @fjaviersanchez Can you point me to where these files are living?  I can try to dig in to it.

**Comment by craiglagegit** [comment]
Hi, weighing in here. There is no allowance for gain in the BF as I have implemented it in GalSim.  I always understood that the images I was dealing with were in electrons.  So if this isn't the case, in other words if the images my code is being presented are in ADU, then the gain change could explain what has changed.  How was the gain change accounted for?  Was the flux of objects at a given magnitude in ADU reduced so that you got the same number of electrons before and after the gain change?

**Comment by fjaviersanchez** [comment]
@cwwalter @jchiang87 @rmjarvis I discovered why the matching was so bad that it looked that everything was "by chance" matching. The x,y directions in the centroid file are flipped with respect to the DM processed catalogs: x in the centroid file matches with `base_SdssShape_y` and y matches with `base_SdssShape_x`. In the example below I am showing with red markers all the objects with flags==12 and all the detected objects with saturated pixels (flipping x,y) and they seem to match well:

![image](https://user-images.githubusercontent.com/13971732/50021859-320e3c00-ff8f-11e8-891e-80ac8c61278f.png)

**Comment by rmjarvis** [comment]
> I always understood that the images I was dealing with were in electrons. 

That is the case.  We build the image in electrons, then at the very end, one step in Jim's afterburner code simulating the readout process is to apply the gain.  So that's not an issue here.

**Comment by fjaviersanchez** [comment]
@rmjarvis the files live here: `/global/cscratch1/sd/jchiang8/desc/Run2.1i/focal_plane_tests/output/rerun/jchiang/w_2018_48/sip_order_2`

**Comment by rmjarvis** [comment]
Thanks.  Can you also point me to the corresponding 2.0i images?  I want to compare the two for the same exposure if possible to see if I can see what relevant difference there might be.

**Comment by jchiang87** [comment]
Unfortunately, we don't have a matching exposure in the Run2.0i data.

**Comment by rmjarvis** [comment]
Oh.  I thought [these plots](https://github.com/LSSTDESC/DC2-production/issues/291#issuecomment-447209299) were comparing the two runs on the same exposure.  No?

**Comment by jchiang87** [comment]
Similar seeing, but different exposures.

**Comment by rmjarvis** [comment]
OK.  Can you point me to which 2.0 exposure that was then?  Maybe I can still find a salient difference without directly matching sources.

**Comment by jchiang87** [comment]
Here are the calexps for that one:  /global/cscratch1/sd/jchiang8/desc/Run2.0i/output_2018-10-20/rerun/jchiang/w_2018_39/calexp/00300187-r

**Comment by jchiang87** [comment]
The raw files are in  /global/cscratch1/sd/jchiang8/desc/Run2.0i/output_2018-10-20/raw/300187
and the 2.1i raws are in /global/cscratch1/sd/jchiang8/desc/Run2.1i/focal_plane_tests/output/raw/1918527

**Comment by rmjarvis** [comment]
OK, so here are some numbers, some of which don't make a lot of sense to me, so I'll just put them out here for comment.

First let's look at the sky level.  As mentioned above, the Run 2.1i image does have a much higher sky level than the 2.0i image.  I think I know which images are in ADU and which are in electrons, and I converted to e- for the ADU images using the respective gains.

| Estimates of sky level | Run 2.0i image | Run 2.1i image
|---: | --- | ---
|Median raw pixel (ADU) | 1537 | 3733
|   (e-) | 1076 | 6346
|Median variance plane (e-) | 362 | 3966
|Median bkgd pixel (ADU) | 532 | 2735
|   (e-) | 372 | 4650

The main thing that I don't understand here is that the raw images both seem to have an extra 1000 ADU above the sky background.  Is that really how high the bias level is set?  That seems a bit excessive.  Or is there some other explanation for this that I'm not thinking of?

Other than that, the only number that seems off is that the variance plane for 2.1i seems about 15% too low.  I think it should be *higher* than the bkgd image converted to e-, since it should be that plus the read noise.  But in both cases, it comes out lower.  (Only slightly for 2.0i.)  So there might be an error in something there.  Or I might not understand what this image is really supposed to mean.

Now, let's see if we can understand what is going on with saturation.

| Estimates of saturation  | Run 2.0i image | Run 2.1i image
|---: | --- | ---
|Maximum calexp pixel (e-) | 52004 | 51936
|Maximum raw pixel (ADU) | 59835 | 18955
|   (e-) | 41884 | 32223
|Maximum variance plane (e-) | 36347 | 82843

The two calexp maxima are similar, which originally I thought was a good sign that saturation was happening at the same pixel value in both places.  It's not 1e5 electrons, as Jim had suggested though.  Even with the very different sky backgrounds, in both cases, the sky is still negligible compared to the maximum, so that doesn't seem terribly relevant.

But even more confusing are the other images.  When I convert the raw pixel values to electrons, they don't match what the calexp came up with for the maximum.  (I thought it should be roughly the calexp maximum plus the sky level.)  Nor are they close to equal, which I thought they should be, since that should have been where Jim cut off the e-image at the saturation level.

And finally, the variance plane really doesn't make sense to me.  They shouldn't be that different I don't think, even with the different sky levels (which contribute 4300 to the difference, an order of magnitude less than the observed difference).

**Comment by jchiang87** [comment]
The raw files are in ADU and I'm pretty sure the calexps are as well (esp. given Robert's statement about not knowing the gains with much certainty).  There are no eimages for these runs, so no pixels are e-.  The bias level is 1000 adu/pixel, so that needs to be subtracted from the raw images. (The bias levels in the real sensors are much higher btw). That means the sky level in the 2.0i image is 537*1.7 = 912 e-/pixel (r-band is nominally 800 e-/pixel for dark skys for 30s exposurs), and for the 2.1i image, 2733*0.7 = 1913 e-/pixel.   The implied saturation levels from those data are for 2.0i, 58835*1.7 = 100019 e-/pixel (check), and for 2.1i, 17955*0.7=12568 e-/pixel (that max raw adu for 2.1i most not be correct).

**Comment by jchiang87** [comment]
In a Run2.1i raw that I just simulated, the max adu value (which is in a bleed trail) is 143860, so the saturation level is 142860*0.7 = 100002. e-/pixel.

I made a raw file for the associated eimage with gain 1.7 (instead of 0.7), and ran processCcd.py on the 0.7 and 1.7 versions.  The brightest star in the 0.7 gain version has r-mag = 18.4, while the 1.7 gain version has r-mag=17.5.   This star is saturated, so I don't think the issue is in the sims, and I'm expecting/hoping this mismatch just occurs for saturated objects.  I'll check that next.

**Comment by jchiang87** [comment]
Here's a plot of measured r-magnitudes from the gain=1.7 `raw` file vs the r-mags for the gain=0.7 `raw` file, generated from the same eimage.  At the bright end, where the saturated stars lie, there is the ~1 mag offset, but aside from a mis-matched source or two, and measurement noise at the faint end, the measured magnitudes agree, as expected.  So I think everything is ok.
![gain1 7_vs_gain0 7_r-mag_comparison](https://user-images.githubusercontent.com/1994473/50032643-17958c00-ffaa-11e8-8095-7b5f073e6faf.png)

**Comment by rmjarvis** [comment]
Update: A few corrections.  First I had the gains reversed, which was part of my confusion.  Jim's post reminded my that 1.7 was the 2.0i gain, and 0.7 is the new gain for 2.1i.  

Second, I was only accessing one of the amps for the raw image, rather than all 16.  So the maximum was spuriously low for 2.1i in particular (where it happened that there wasn't a saturated pixel in the one amp I grabbed).  

So let me update my numbers:

Estimates of sky level | Run 2.0i image | Run 2.1i image
--: | -- | --
Median raw pixel (ADU) | 1537 | 3735
(e-) | 2611 | 2614
Median bkgd pixel (ADU) | 532 | 2735
(e-) | 904 | 1915
Median variance * gain (ADU) | 616 | 2776 

So now those all make sense, given a bias level of ~1000 ADU.  This means the sky in this 2.1i image is only 2x the 2.0i image, which seems more reasonable.  And the variance plane makes sense as being somewhat more than the variance from the sky level, since there would be other sources of noise (bias, dark) as well.

Furthermore, the actual pixel variance of the (roughly) non-signal pixels in the calexp matches the value in the variance plane reasonably well (554, 2759), so I think that confirms that the calexp is indeed in ADU. I assumed they would have multiplied by the gain to get e-, but I guess not.

Moving on to the saturation level...

Estimates of saturation | Run 2.0i image | Run 2.1i image
--: | -- | --
Maximum calexp pixel (ADU) | 52004 | 51936
(e-) | 88408 | 36355
Maximum raw pixel (ADU) | 59835 | 144030
(e-) | 101719 | 100821
Maximum variance plane * gain (ADU) | 61789 | 57990

So both sims now have maximum raw pixel values close to 1.e5 electrons, as Jim said.  So that makes sense.  The maximum pixel in the calexp for 2.0i is a little lower, but this is probably because it is interpolating the pixels that were saturating, and underestimating a bit.  The maximum implied by the variance plane is a bit closer, and probably makes sense, but I didn't try to track it through.

But the calexp for 2.1i seems to have much too low a maximum.  And the variance plane's maximum is also much too low.  My hypothesis for what is going on here is that there is a value being given to DM to tell it what the saturation level is, but we didn't update that when we changed the gain.

It should have been 58823 (1.e5/1.7) in run 2.0i.
It should be 142857 (1.e5/0.7) in run 2.1i.

I think this explains Jim's mag-mag plot getting discrepant at high magnitudes.  And also Javi's plot of the bright (but not saturated) star sizes being messed up.

**Comment by jchiang87** [comment]
Great! Your hypothesis regarding the saturation level is correct:  There's a parameter I wasn't aware of in the obs_lsst policy files:

https://github.com/lsst/obs_lsst/blob/master/policy/cameraHeader.yaml#L58

This also explains Roberts comment about saturation being in ADU that seemed off-point to me before.
We'll need to fix that in obs_lsst before processing the Run2.1i data.

**Comment by cwwalter** [comment]
Awesome! In  2.0i were we not using this package so it wasn't an issue? Can we reprocess the run we already have to see if we can see the BF again?

Also, what exactly is DM doing with the saturation number we give it?

**Comment by jchiang87** [comment]
We were using that package for Run2.0i, but the saturation value (56000*1.7=95200 e-/pixel) is not far off from the right value.  We can reprocess the Run2.1i data with the right saturation value in the policy file, but I don't think it will change the B/F part of the curve for that visit with poor seeing.

**Comment by rmjarvis** [comment]
I think DM is masking every pixel with a value greater than this nominal saturation value.  And then for the calexp, it interpolates all masked pixels (I wish they wouldn't, but don't get me started on that!).  Since the pixels around those that are marked saturated are lower than the masked ones, they interpolate to somewhere near, but lower than, the nominal saturation level.

For the bright-but-not-saturated stars, this ends up basically cutting off their peak.  So they are measured to be larger than they should be.  (They lost flux near 0,0, so the second moments get pushed out.)  This is the steep size climb characteristic of saturated stars.

But this means the brightest object that doesn't look saturated can only have a peak value of about 36,000.  Way lower than the 100,000 that we should get.  Those objects don't have much of a B/F effect, so it's hard to see it in the 2.1i plot.  If we stopped cutting off their centers, I think we will probably see the B/F effect back again.

**Comment by jchiang87** [comment]
Setting `saturation = 142857` in the obs_lsst policy file and re-analyzing the special-purpose sensor-visit I made, here is the gain=0.7/1.7 comparison (plotting delta-mag vs mag):
![gain1 7_vs_gain0 7_r-mag_comparison](https://user-images.githubusercontent.com/1994473/50035608-0c495d00-ffb8-11e8-924c-4a66677cc0bc.png)

**Comment by cwwalter** [comment]
I understand the scatter at the low end.  Any idea why you get the outliers for the bright guys?

**Comment by jchiang87** [comment]
I haven't looked into it, but I'm just doing positional matching, using a 0.1 arcsec radius, so they may be mis-matches.

**Comment by rmjarvis** [comment]
I suspect it is that different pixels get tagged as saturated, since the 1.7 version didn't have the saturation level at 1.e5/1.7.  It was a bit lower.  So the precise pixels that are masked and then interpolated over will be different.  The resulting magnitudes would be close, and probably statistically similar, but not precisely equal.

**Comment by fjaviersanchez** [comment]
@jchiang87 after your fix things are looking good:

![image](https://user-images.githubusercontent.com/13971732/50039594-ad9cd700-ffe9-11e8-877d-9a5a141f84b4.png)

zooming in:

![image](https://user-images.githubusercontent.com/13971732/50039601-be4d4d00-ffe9-11e8-9428-444b3a655100.png)

Thanks a bunch!

**Comment by cwwalter** [comment]
That's great. We can see the BF again. 

OK I think if @fjaviersanchez is OK on the knot checks then all of our planned checks have now been completed.

**Comment by cwwalter** [comment]
Hmm... I think I recall Francois did this by differencing against an exposure without them, which we can't do here.

@EiffL Do you have any suggestions for checking if the knots are as expected in single exposures?

**Comment by EiffL** [comment]
Thanks Javier for this, I suspect it's going to be very difficult to see the knots. We manually restricted the flux that goes into the knots as a function of size, so for large galaxies where we could potentially visually resolve individual knots, they are probably (by design) to faint to see. 
The best test I can think of if we want to check that they are actually on, is to run a given visit on a modified instance catalog where they are explicitly removed and look at the difference.

**Comment by rmjarvis** [comment]
They look definitely on to me.  Those images that Javi posted are much chunkier than you would expect from just Poisson photon noise.  There are knots in those profiles, so I don't think there's any concern that we accidentally didn't turn on the knots.  They look fine to me.

**Comment by fjaviersanchez** [comment]
@sethdigel made some preliminary analyses on the 2.1i files. I am putting a link to the messages at slack in this issue to keep track of this: https://lsstc.slack.com/archives/C77DDKZHR/p1546297981015900

He found some FITS files with zero-size centroid file associated to them. I agree with him that this is probably not worrisome but we should try to investigate what's the cause (unfinished jobs, I/O problems, other?)

Thanks a lot Seth!

**Comment by sethdigel** [comment]
> He found some FITS files with zero-size centroid file associated to them. I agree with him that this is probably not worrisome but we should try to investigate what's the cause (unfinished jobs, I/O problems, other?)

I sent a list of the 92 empty centroid files (each of which has a corresponding FITS image) to Antonio in case it is useful for tracking down the issue.
